backend/server/routers/ctf.py

```diff
@@ -0,0 +1,283 @@
+    }
+
+
+def get_code(course: str) -> int:
+    """
+    Returns the code of a courseCode
+    EG: COMP1511 -> 1511
+    """
+    return int(course[4:])
+
+
+def get_faculty(course: str) -> str:
+    """
+    Returns the faculty of a courseCode
+    EG: COMP1511 -> COMP
+    """
+    return course[:4]
+
+
+def gen_eds(courses: set[str]) -> set[str]:
+    """
+    Returns all gen eds from a set of courses
+    """
+    return set(
+        course
+        for course in courses
+        if not course.startswith("COMP") and not course.startswith("MATH")
+    )
+
+
+def hard_requirements(data: PlannerData) -> bool:
+    # NOTE: Can't check start year from this
+    # Frontend should handle most of this anyways
+    # including validity of the program
+    return (
+        data.program == "3778"
+        and "COMPA1" in data.specialisations
+        and "MATHC2" in data.specialisations
+        and len(data.plan) == 3
+    )
+
+
+def extended_courses(data: PlannerData) -> bool:
+    """
+    Must take atleast 3 courses with extended in the name
+    """
+    extended_courses = {
+        "COMP3821",
+        "COMP3891",
+        "COMP6841",
+        "COMP6843"
+        "COMP6845"
+    }
+    return len(extended_courses & all_courses(data)) >= 3
+
+
+def summer_course(data: PlannerData) -> bool:
+    """
+    Must take atleast one summer course
+    """
+    return any(
+        course.startswith("COMP")
+        for year in data.plan
+        for course in year[0]
+    )
+
+
+
+def term_sums_even(data: PlannerData) -> bool:
+    """
+    Check that the sum of the course codes in even terms is even
+    """
+    is_even: Callable[[int], bool] = lambda x: x % 2 == 0
+    print("Checking even")
+    for y, year in enumerate(data.plan):
+        # Exclude summer term + odd terms
+        for i, term in enumerate(year[2::2], 2):
+            term_sum = sum(map(get_code, term.keys()))
+            print(f"{y}T{i} sum: {term_sum}")
+            if not is_even(term_sum):
+                print("failed: ", term)
+                return False
+
+    return True
+
+# TODO
+def term_sums_odd(data: PlannerData) -> bool:
+    """
+    Check that the sum of the course codes in odd terms is odd
+    """
+    is_odd: Callable[[int], bool] = lambda x: x % 2 == 1
+    print("Checking odd")
+    for year in data.plan[::2]:
+        # Exclude summer term + even terms
+        for term in year[1::2]:
+            term_sum = sum(map(get_code, term.keys()))
+            if not is_odd(term_sum):
+                print("failed: ", term)
+                return False
+    return True
+ 
+def comp1511_marks(data: PlannerData) -> bool:
+    """
+    Ollie must achieve a mark of 100 in COMP1511 to keep his scholarship
+    """
+    for year in data.plan:
+        for term in year:
+            for course in term:
+                _, marks = term[course]  # type: ignore
+                if course == "COMP1511":
+                    return marks == 100
+
+    return False
```

Comment:
thoughts on turning this into a generator comprehension with `next`?

Reply:
true, this can be triv-solved with
```py
return any(
  course
  for term in year
  for course, mark in term
  if course == "COMP1511" and mark == 100
)
```

Reply:
true don't even need next, but what about simplifying this even further and doing
```python
return any(
  course == "COMP1511" and mark == 100
  for term in year
  for course, mark in term
)
```

Reply:
THIS is incredibel +++EV!!!!!!